src/streamsync/core_ui.py

```diff
@@ -0,0 +1,141 @@
+        return self.components.get(component_id)
+
+    def get_direct_descendents(self, parent_id: str) -> List[Component]:
+        children = list(filter(lambda c: c.parentId == parent_id,
+                               self.components.values()))
+        return children
+
+    def get_direct_descendents_length(self, parent_id):
+        return len(self.get_direct_descendents(parent_id))
+
+    def get_descendents(self, parent_id: str) -> List[Component]:
+        children = self.get_direct_descendents(parent_id)
+        desc = children.copy()
+        for child in children:
+            desc += self.get_descendents(child.id)
+
+        return desc
+
+    def attach(self, component: Component, override=False) -> None:
+        self.counter += 1
+        if (component.id in self.components) and (override is False):
+            raise RuntimeWarning(f"Component with ID {component.id} already exists")
+        self.components[component.id] = component
+
+    def ingest(self, serialised_components: Dict[str, Any]) -> None:
+        removed_ids = self.components.keys() - serialised_components.keys()
+
+        for component_id in removed_ids:
+            if component_id == "root":
+                continue
+            self.components.pop(component_id)
+        for component_id, sc in serialised_components.items():
+            component = Component(**sc)
+            self.components[component_id] = component
+
+    def to_dict(self) -> Dict:
+        active_components = {}
+        for id, component in self.components.items():
+            active_components[id] = component.to_dict()
+        return active_components
+
+
+class SessionComponentTree(ComponentTree):
+
+    def __init__(self, base_component_tree: ComponentTree):
+        super().__init__()
+        self.base_component_tree = base_component_tree
+
+    def get_component(self, component_id: str) -> Optional[Component]:
+        # Check if session component tree contains requested key
+        session_component_present = component_id in self.components
+
+        if session_component_present:
+            # If present, return session component (even if it's None)
+            session_component = self.components.get(component_id)
+            return session_component
+
+        # Otherwise, try to obtain the base tree component
+        return self.base_component_tree.get_component(component_id)
+
+    def attach(
+            self,
+            component: Component,
+            override=False,
+            attach_to_bottom=False) -> None:
+        if attach_to_bottom is True and component.position == 0:
+            if component.parentId is not None:
+                tree = self \
+                        if component.parentId in self.components \
+                        else self.base_component_tree
+                parent_length = \
+                    tree.get_direct_descendents_length(component.parentId)
+                component.position = parent_length + 1
```

Comment:
To get the position you should check the maximum position, then add 1.

We need to account for positionless components (only _Sidebar_ at this point)

Creation of _Sidebar_ should also hardcode position to `-2`. This is a special value that indicates that it should be rendered in the positionless slot.

Reply:
Actually, check out `getNextInsertionPosition()` in `useComponentActions.ts`.

If there are children, should be the 
Math.max(max position + 1, 0)

No children:
0